rdrf/rdrf/auth/views.py

```diff
@@ -0,0 +1,175 @@
+
+    if user_needs_verification():
+        user_data = _get_users_verification_data(user)
+
+        def verification_page(form=None):
+            context = {
+                'title': _('Verify User'),
+                'validlink': True,
+                'form': form,
+            }
+            return TemplateResponse(request, 'registration/login_assistance_verify_user.html', context)
+
+        if request.method != 'POST':
+            if user_data is None:
+                context = {
+                    'title':  _("Can't verify user")
+                }
+                template_name = 'registration/login_assistance_can_not_verify.html'
+                return TemplateResponse(request, template_name, context)
+
+            return verification_page()
+
+        else:
+            # Verify the data POSTed by the user
+            # In case the verification succeeds we save the token in the session, so that the user can progress to
+            # the next phase by just revisiting the same page
+            form = UserVerificationForm(user_data, request.POST)
+            if form.is_valid():
+                request.session['user_verified'] = {
+                    'uidb64': uidb64,
+                    'token': token,
+                }
+                # Redirect to the same page for the next phase
+                return HttpResponseRedirect(reverse('login_assistance_confirm',
+                    kwargs={'uidb64': uidb64, 'token': token}))
+            else:
+                # User verification failed, display same page with errors
+                return verification_page(form)
+
+    # The user has been verified now.
+    # We will display the reason why they have been locked out and require them to set a new password if
+    # their password expired, or allow them to optionally set a new password in all other cases.
+
+    deactivation = UserDeactivation.objects.filter(username=user.username).first()
+    deactivation_reason = deactivation.reason if deactivation is not None else None
+
+    DEFAULT_REASON_TEXT = _('Your account is currently locked.')
+    REASON_TEXTS = {
+        UserDeactivation.ACCOUNT_EXPIRED:
+            _('Your account is suspended because your account was inactive for too long.'),
+        UserDeactivation.PASSWORD_EXPIRED:
+            _('Your account is suspended because your password expired.'),
+        UserDeactivation.TOO_MANY_FAILED_LOGINS:
+            _('Your account is suspended because you (or someone else) entered an incorrect password too many times.')
+    }
+    reason = REASON_TEXTS.get(deactivation_reason, DEFAULT_REASON_TEXT)
+
+    def is_password_change_required(user):
+        # If the password has expired always make them change their password or they will
+        # be locked out when they try to log in next time.
+        return deactivation_reason == UserDeactivation.PASSWORD_EXPIRED or is_password_expired(user)
```

Comment:
could there be a userdeactivation object with reason UserDeactivation.PASSWORD_EXPIRED but is_password_expired(user)  false?

Reply:
Yes. `UserDeactivation.PASSWORD_EXPIRY` is set by `django-useraudit` only if `django-useraudit` deactivated the user account and because of password expiry.

The `is_password_expired` check is there for cases when the user has been deactivated for other reasons, but their password is also expired.

If we remove that check and allow them to reactivate their account without changing their password, they will get locked out because of their password expired when they will try to log in.